Improve performance of accessing per-particle energy, forces, torques, and virials properties

tcmoore3 opened the pull request:
## Description

Use the GatherTagOrder class to properly gather the per-particle properties from all ranks instead of the slow, previous implementation. This change should drastically improve the performance of (benchmarks to come):
* ForceCompute::getEnergiesPython()
* ForceCompute::getForcesPython()
* ForceCompute::getTorquesPython()
* ForceCompute::getVirialsPython()

I added a `vec6` struct for the `getVirialsPython()` function since you cannot create vectors of arrays like `std::vector<double[6]>`. 

## Motivation and context

Resolves #1539.

## How has this been tested?

Updated tests in `md/pytest/test_custom_force.py` that sets the properties of a `Custom` force to be functions of the particle tag, so that you can check that the returned values are returned in the correct order.

## Change log

<!-- Propose a change log entry. -->
```
Improve performance of accessing per-particle properties of `Force` objects.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

tcmoore3 committed:
Use GatherTagOrder to gather the global energy array

tcmoore3 committed:
Use proper gather to sort forces, torques, and virials

I added a vec6 Struct to hold the virials since you cannot create
Vector-s of arrays. I also updated md/pytest/test_custom_force.py to set
the values of the local force arrays to be functions of the particle tag
to test that the values are returned in the proper order.

joaander commented on hoomd/ForceCompute.h:
```diff
@@ -64,6 +65,94 @@ class PYBIND11_EXPORT ForceCompute : public Compute
+                    "Attempting to access non-existent vec6 entry (i.e. i > 2)");
+#endif
+                }
+            }
```
```suggestion
```
This code is not used.

joaander commented on hoomd/md/pytest/test_custom_force.py:
```diff
@@ -115,6 +115,51 @@ def test_simulation(local_force_names, force_simulation_factory,
+    def set_forces(self, timestep):
+        with self.cpu_local_force_arrays as force_arrays:
+            with self._state.cpu_local_snapshot as local_snapshot:
+                force_arrays.force[:] = -5
```
Need a non-trivial force vector.

joaander commented on hoomd/md/pytest/test_custom_force.py:
```diff
@@ -115,6 +115,51 @@ def test_simulation(local_force_names, force_simulation_factory,
+                force_arrays.force[:] = -5
+                energy = local_snapshot.particles.tag * -10
+                force_arrays.potential_energy[:] = energy
+                force_arrays.torque[:] = np.ones(
```
We should test values of forces, torques, and virials where all the components are different values. This verifies e.g. that the proper ordering of the virial components is maintained through all the steps of the sorting operation.

Something like:
```
force = numpy.stack(
    (local_snapshot.particles.tag * 3, local_snapshot.particles.tag * 5, local_snapshot.particles.tag * -7), axis=-1)
```

joaander commented on hoomd/ForceCompute.h:
```diff
@@ -64,6 +65,94 @@ class PYBIND11_EXPORT ForceCompute : public Compute
         m_deltaT = dt;
         }
 
+    template<class Real> struct vec6
```
```suggestion
    namespace detail {
    template<class Real> struct vec6
```
And a closing brace at the end of the class. This makes it clear that the class is a detail of the implementation and not part of the public API.

joaander reviewed (changes requested):
Thanks! I have some suggestions to improve this.

tcmoore3 committed:
Revert "Use proper gather to sort forces, torques, and virials"

This reverts commit e633cdf9087c9137a83687312d56d4a15952fcbc.

tcmoore3 committed:
Add vec6 to hoomd::detail

b-butler commented on hoomd/ForceCompute.cc:
```diff
@@ -336,26 +343,39 @@ pybind11::object ForceCompute::getEnergiesPython()
     else
         {
-        return pybind11::none();
+        global_energy = local_energy;
```
```suggestion
        global_energy = std::move(local_energy);
```
I think this will perform a copy? Should we do an explicit `std::move`?

b-butler commented on hoomd/ForceCompute.cc:
```diff
@@ -336,26 +343,39 @@ pybind11::object ForceCompute::getEnergiesPython()
+    // sort energies by particle tag
+    std::vector<double> local_energy;
+    local_energy.reserve(m_pdata->getN());
+    std::vector<uint32_t> local_tag(m_pdata->getN());
```
Might be worth a helper function to get sorted local tags. We use the same code in each of the functions.

b-butler commented on hoomd/ForceCompute.cc:
```diff
@@ -377,23 +397,38 @@ pybind11::object ForceCompute::getForcesPython()
+        }
+    else
+        {
+        global_force = local_force;
```
```suggestion
        global_force = std::move(local_force);
```
See previous comment about `std::move`.

b-butler commented on hoomd/ForceCompute.cc:
```diff
@@ -420,23 +455,38 @@ pybind11::object ForceCompute::getTorquesPython()
+        }
+    else
+        {
+        global_torque = local_torque;
```
```suggestion
        global_torque = std::move(local_torque);
```

b-butler commented on hoomd/ForceCompute.cc:
```diff
@@ -468,32 +518,41 @@ pybind11::object ForceCompute::getVirialsPython()
+        }
+    else
+        {
+        global_virial = local_virial;
```
```suggestion
        global_virial = std::move(local_virial);
```

b-butler reviewed (changes requested):
@tcmoore3 Thanks for the changes, I have a few suggestions, but otherwise looks good.

tcmoore3 committed:
Fix comment

tcmoore3 committed:
Update tests

tcmoore3 committed:
Use std::move as suggested in code review

Co-authored-by: Brandon Butler <butlerbr@umich.edu>

tcmoore3 committed:
Add getSortedLocalTags helper function

Every get*Python() function in ForceCompute does the same thing, so this
commit adds a helper function to call in each of those functions. I
implemented the function and use it in getEnergiesPython() here, and it
compiles and passes tests. Next commit will update the other
get*Python() functions to use getSortedLocalTags().

tcmoore3 committed:
Rename and use ForceCompute::getSortedLocalTags

I renamed the helper function from getSortedLocalTags to sortLocalTags
because the latter more accurately describes what the function does.

I also now use the function in all of the ForceCompute::get*Python()
functions. All python tests, serial and parallel, are passing.

joaander reviewed (approved):
Thanks!

b-butler reviewed (approved):
Thanks @tcmoore3!
